### ISSUE TITLE
NPE Attempt to invoke virtual method int com.couchbase.lite.Database.findMissingRevisions on a null object reference

### ISSUE THREAD

**Issue by jamiltz** [couchbase/couchbase-lite-java-core#1015]
Not sure how to reproduce this exception but I got it when the connectivity was low.

```
E/Sync: ERROR: processInbox failed: 
   java.lang.NullPointerException: Attempt to invoke virtual method 'int com.couchbase.lite.Database.findMissingRevisions(com.couchbase.lite.RevisionList)' on a null object reference
       at com.couchbase.lite.replicator.PullerInternal.processInbox(PullerInternal.java:178)
       at com.couchbase.lite.replicator.ReplicationInternal$2.process(ReplicationInternal.java:314)
       at com.couchbase.lite.support.Batcher.processNow(Batcher.java:372)
       at com.couchbase.lite.support.Batcher.access$300(Batcher.java:16)
       at com.couchbase.lite.support.Batcher$2.run(Batcher.java:306)
       at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:422)
       at java.util.concurrent.FutureTask.run(FutureTask.java:237)
       at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:152)
       at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:265)
       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
       at java.lang.Thread.run(Thread.java:818)
```

Will try to run more tests to reproduce.

**Comment by jamiltz** [comment]
Note: From that point, I notice that the `/_active_tasks` endpoint doesn't return the replicator (of type push) anymore but documents are are being synched. Every time a change on the local db occurs the exception above is thrown.

**Comment by jamiltz** [comment]
To reproduce:

Start this app (https://github.com/jamiltz/P2PSync-Android) and check out on the issue/npe_findMissingRevisions branch.
- Use two phones connected to same wifi
- On device 1, click cabin
- On device 2, click galley
- The total number of docs on each device should be equal
- Take device 2 on the _edge_ of the network
- Bring back device 2 in range

The replications still occur but the `_active_tasks` only returns one replicator and and the live queries only fires when local changes are made.

**Comment by pasin** [comment]
The exception is the exact same thing as what I have commented in there
https://github.com/couchbase/couchbase-lite-java-core/issues/976#issuecomment-179706531

**Comment by hideki** [comment]
this is duplicate with #1026

**Comment by NitzDKoder** [comment]
@hideki had the same exception on low - cellular network..

from 
java.lang.NullPointerException: Attempt to invoke virtual method 'int com.couchbase.lite.Database.findMissingRevisions(com.couchbase.lite.RevisionList)' on a null object reference
	at com.couchbase.lite.replicator.PullerInternal.processInbox(SourceFile:163)
to

java.lang.RuntimeException: java.lang.NullPointerException: Attempt to invoke virtual method 'int com.couchbase.lite.Database.findMissingRevisions(com.couchbase.lite.RevisionList)' on a null object reference

Is this fixed?

@pasinpasin closed this in fc011c3 4 days ago
 @pasinpasin removed the in progress label 4 days ago

Planning to take changes top of 1.2 release can you help on that?

thanks
Nithin

**Comment by hideki** [comment]
Hi @NitzDKoder 

This issue was recently fixed. Could you please try the build from master branch?

maven repo
http://files.couchbase.com/maven2/com/couchbase/lite/couchbase-lite-android/0.0.0-581/

Zip download
http://latestbuilds.hq.couchbase.com/couchbase-lite-android/0.0.0/0.0.0-581/

Thanks!
Hideki

**Comment by joshblour** [comment]
Hey @hideki,

I'm getting this error, even after I've upgraded to `com.couchbase.lite:couchbase-lite-android:0.0.0-607`

**Comment by hideki** [comment]
Hi @joshblour,

Could you please provide log message or stacktrace? If you have other information, please provide these too.

Thanks,
Hideki

**Comment by joshblour** [comment]
sure! here you go: https://gist.github.com/joshblour/0376ea11b077c55fb119

**Comment by hideki** [comment]
Reopen this ticket: There is scenario that pull replicator is running after database is closed. Need to investigate.

**Comment by hideki** [comment]
https://gist.github.com/joshblour/0376ea11b077c55fb119 is not accessible anymore.

I believe this is fixed with v1.2.1. I will close this ticket.
